### DISCUSSION TITLE
farm power output negative and wrong for high wind speeds

### DISCUSSION THREAD

**Discussion by henrysanders3** [NatLabRockies/floris#583] [General]
Hi,

After optimising the yaw angles and plotting the power output for the farm, the power is higher than is physically possible for the turbine types and some are plotting as negative.

Would be great to understand the causes of this and the changes required.

many thanks

**Comment by paulf81** [comment]
hi, can you share a small example that produces the error that we can use to investigate?  thank you!

**Reply by henrysanders3** [reply]
Hi, 

below is a screenshot of the code run. at wind speeds of 24 m/s the code plots negative values for the power output.

![image](https://user-images.githubusercontent.com/78905038/220414038-1addd99f-e083-487f-a40d-09b9e545ad6b.png)

![image](https://user-images.githubusercontent.com/78905038/220414154-bbd7afcf-0da9-4701-bc57-acf4086524a7.png)


Many thanks for your help!

**Reply by paulf81** [reply]
Thanks for posting this, I think I know what is going on here.  At 24 m/s the power curve is not perfectly flat wrt wind speed as it should be (a live issue (Issue #554)), as a result turbines can yaw to lower their effective wind speed, raising the power by 0.002%, which is not physical.  The solution will be to fix #554  which is on the todo list, in the meanwhile:

@Bartdoekemeijer : We could increase the lower bound for verify_convergence?

@rafmudaf  just flagging this discussion to see what you think.

**Reply by cherifM** [reply]
Hi, I think im having the same issue  even im doing layout optimization (not yaw optimization), I was thinking to add "try and raise exception" without stopping the iterations,  in file file "floris/simulation/turbine.py", , something like this :
![image](https://user-images.githubusercontent.com/7900713/220569659-3e205000-571d-4c9a-bd01-5396fd144dc8.png)
or maybe we can check if power is not physically correct, than return p=0? and move to the next iteration
what you think? verify_convergence is available only in yaw_optimization mode, so i cant set this in layout_optimization files.
cherif

**Reply by Bartdoekemeijer** [reply]
> @Bartdoekemeijer : We could increase the lower bound for verify_convergence?

Yes, that should definitely solve it. Though that's really more of a bandage than a solution.

**Reply by cherifM** [reply]
@Bartdoekemeijer 
hi,  verify_convergence is available only in yaw_optimization mode|clases, so I cant set this in layout_optimization files.
im having the same issue and i cant do even "bandage" solution , at the moment im using  "try and exception":
![image](https://user-images.githubusercontent.com/7900713/220634416-e0bbef8c-e539-423d-8895-dae6966652d5.png)
please any idea or solution ?
best regards

**Reply by rafmudaf** [reply]
Just as a sanity check - @henrysanders3 are you using the power curve from the NREL 5MW turbine? I see that rotor diameter is 107 meters, so I just want to verify that your power curve looks like the one @paulf81 showed in #554 and that the oscillations are indeed the issue.
Also, it's helpful to have your code directly so that we can copy it and run it locally to debug. You can replace the screenshots of the code above with the text directly and you can format it as Python by wrapping it in three ticks - that's the character below `~` on a US keyboard. It should look like this:

````
```python
def function():
    pass

a = 1
b = function()
```
````

**Reply by cherifM** [reply]
@rafmudaf   
I guess you right, i plot my ct and it look like it  have issue
![image](https://user-images.githubusercontent.com/7900713/220857573-516d6c5a-edec-402a-892b-3d7bfd1950c3.png)

**Reply by henrysanders3** [reply]
Hi, Below is the code used to optimise the yaw angles

```python
import numpy as np
import matplotlib.pyplot as plt
from floris.tools import FlorisInterface
from floris.tools.optimization.yaw_optimization.yaw_optimizer_sr import (
    YawOptimizationSR,
)
import pandas as pd

# Load the default example floris object
fi = FlorisInterface("inputs/gch.yaml") # GCH model matched to the default "legacy_gauss" of V2
# fi = FlorisInterface("inputs/cc.yaml") # New CumulativeCurl model

# Reinitialize as a 25-turbine farm with range of WDs and 1 WS
D = 107.0 # Rotor diameter 

fi.reinitialize(

    layout_x=[0, -D*3.83, -D*7.76, -D*11.78, -D*15.8, -D*19.81, -D*23.83, -D*27.94, -D*31.96, -D*28.6, -D*24.21, -D*19.81, -D*15.42, -D*6.64, -D*2.24, D*2.15, D*6.54, D*10.93, D*15.33, D*18.69, D*9.81, D*5.42, D*1.03, -D*3.36, -D*7.76],
    layout_y=[0, D*1.4, D*2.9, D*4.49, D*5.98, D*7.48, D*8.97, D*10.97, D*11.87, D*20.75, D*19.07, D*17.38, D*15.7, D*12.34, D*10.65, D*8.97, D*7.29, D*5.61, D*3.93, D*12.8, D*16.17, D*17.85, D*19.53, D*21.21, D*22.90],
    wind_directions=np.arange(100, 250, 15), 
    wind_speeds=[24],
)

print(fi.floris.farm.rotor_diameters)

# Initialize optimizer object and run optimization using the Serial-Refine method
yaw_opt = YawOptimizationSR(fi)#, exploit_layout_symmetry=False)
df_opt = yaw_opt.optimize()

fi.calculate_wake()

print("Optimization results:")
print(df_opt)

# Split out the turbine results
for t in range(25):
    df_opt['t%d' % t] = df_opt.yaw_angles_opt.apply(lambda x: x[t])

# Show the results
fig, axarr = plt.subplots(2,1,sharex=True,sharey=False,figsize=(8,8))

# Yaw results
ax = axarr[0]
for t in range(25):
    ax.plot(df_opt.wind_direction,df_opt['t%d' % t],label='t%d' % t)
ax.set_ylabel('Yaw Offset (deg')
ax.legend()
ax.grid(True)

# Power results
ax = axarr[1]
ax.plot(df_opt.wind_direction,df_opt.farm_power_baseline,color='k',label='Baseline Farm Power')
ax.plot(df_opt.wind_direction,df_opt.farm_power_opt,color='r',label='Optimised Farm Power')
ax.set_ylabel('Power (W)')
ax.set_xlabel('Wind Direction (deg)')
ax.legend()
ax.grid(True)

plt.show()

for t in range(25):
    fig, ax = plt.subplots()
    ax.plot(df_opt.wind_direction, df_opt['t%d' % t], label='t%d' % t)
    ax.set_ylabel('Yaw Offset (deg)')
    ax.legend()
    ax.grid(True)
    


fig1, ax1 = plt.subplots(figsize=(8, 8))
for t in range(25):
    ax1.plot(df_opt.wind_direction, df_opt['t%d' % t], label='t%d' % t)
ax1.set_ylabel('Yaw Offset (deg)')
ax1.legend()
ax1.grid(True)

# Create the second figure and subplot for the power results
fig2, ax2 = plt.subplots(figsize=(8, 8))
ax2.plot(df_opt.wind_direction, df_opt.farm_power_baseline, color='k', label='Baseline Farm Power')
ax2.plot(df_opt.wind_direction, df_opt.farm_power_opt, color='r', label='Optimized Farm Power')
ax2.set_ylabel('Power (W)')
ax2.set_xlabel('Wind Direction (deg)')
ax2.legend()
ax2.grid(True)

# Show the figures
plt.show()

##################################################################


# Print the optimised yaw angles
print("Optimized yaw angles:", df_opt.yaw_angles_opt)

print("Optimization results:")
print(df_opt)
###################################################################
for t in range(25):
    df_opt['t%d' % t] = df_opt.yaw_angles_opt.apply(lambda x: x[t])
    
    
############################################################################
# Baseline and Optimised AEP's
# Add columns for baseline and optimized AEP
df_opt['baseline_AEP'] = df_opt.apply(lambda row: row.farm_power_baseline * 8760, axis=1)
df_opt['optimized_AEP'] = df_opt.apply(lambda row: row.farm_power_opt * 8760, axis=1)

# Print the results
print(df_opt[['wind_direction', 'baseline_AEP', 'optimized_AEP']])

############################################################################
import openpyxl

# Create a new Excel workbook
workbook = openpyxl.Workbook()

# Add a sheet to the workbook
sheet = workbook.create_sheet("AEP Table")

# Add the column headings
sheet.cell(row=1, column=1).value = "Wind Direction"
sheet.cell(row=1, column=2).value = "Baseline AEP"
sheet.cell(row=1, column=3).value = "Optimized AEP"

# Add the data to the sheet
for i, row in df_opt.iterrows():
    sheet.cell(row=i+2, column=1).value = row.wind_direction
    sheet.cell(row=i+2, column=2).value = row.baseline_AEP
    sheet.cell(row=i+2, column=3).value = row.optimized_AEP

# Save the workbook
workbook.save("aep_table.xlsx")
####################################################################

# Calculate percentage increase for each wind direction
df_opt['percent_increase'] = (df_opt.farm_power_opt - df_opt.farm_power_baseline) / df_opt.farm_power_baseline * 100

# Plot the percentage increase
fig2, ax2 = plt.subplots()
ax2.plot(df_opt.wind_direction, df_opt.percent_increase, label='Percent Increase')
ax2.set_ylabel('Percent Increase (%)')
ax2.set_xlabel('Wind Direction (deg)')
ax2.legend()
ax2.grid(True)

############################################

# Power results

# Plot the graph
fig1, ax1 = plt.subplots()
ax1.plot(df_opt['wind_direction'], df_opt['baseline_AEP'], label='Baseline AEP')
ax1.plot(df_opt['wind_direction'], df_opt['optimized_AEP'], label='Optimised AEP')


ax1.set_xlabel('Wind Direction (degrees)')
ax1.set_ylabel('Annual Energy Production (kWh)')
ax1.legend()

plt.show()
```

I have used a different power curve to that of the NREL 5MW, below is the plotted power and thrust curves: 
![image](https://user-images.githubusercontent.com/78905038/220871343-ed9f017d-6717-4e9e-9a86-a48bf034d0b1.png)

**Reply by cherifM** [reply]
@henrysanders3  
as mentioned by  https://github.com/NREL/floris/issues/554#issuecomment-1440697922
you can to edit your ct input data (red line), especially from  ws from 0 to  8 m/s and make it line without any oscilation,  
in my case ,when i do layout optimization (no yaw optim), works just fine, im not getting this error any more,  
hope this help, 
best regards,
cherif

**Comment by paulf81** [comment]
I think we agree that to resolve these issues we'll need to resolve Issue https://github.com/NREL/floris/issues/554 , I'll add some ideas on that issue page

**Comment by Bartdoekemeijer** [comment]
Also worth noting, @henrysanders3, you are optimizing the yaw angles at an ambient wind speed of 24 m/s. Indeed, this leads to numerical issues that should not be there. However, nobody will ever yaw a wind turbine at an ambient wind speed of 24 m/s, both because of the turbine loads and because the wakes are so small that there is nothing realistically to be gained.

**Comment by paulf81** [comment]
This issue is resolved (in v4) by https://github.com/NREL/floris/pull/784
